### ISSUE TITLE
To avoid default eps ineffective in some precision(float16), dynamically set the default value of eps in optimizers such as Adam based on precision.

### ISSUE THREAD

**Issue by 1994cxy** [pytorch/pytorch#147437]
### 🚀 The feature, motivation and pitch

Recently, I have been developing my own model based on the BLIP-2 model, attempting to incorporate some new features into the input of the Q-former. Since the feature extraction network for the new features consumes GPU memory, and my GPU only has 80GB of memory, I used float16 precision to train my model. Additionally, I fixed all the parameters of the BLIP-2 model and only trained my new feature extraction network. 

However, after calling optimizer.step() in the first iteration, I noticed that the parameters of the feature extraction network all became NaN. Initially, I thought it might be due to gradient explosion or an excessively large learning rate. However, I later found that these were not the issues. The problem lay in the second moment of the gradients, which were too small. This caused the sqrt(v_t) in the denominator of the Adam optimizer to become zero. Although PyTorch adds eps to the denominator to prevent division by zero, I discovered during debugging that the default eps=1e-8 does not work effectively under float16 precision. 

I hope that future versions of PyTorch can automatically adjust related parameters based on the precision used for training the model, avoiding issues caused by eps becoming ineffective.

**Comment by janeyx99** [comment]
Do you have ideas on a good policy for how to adjust eps based on the dtype?

**Comment by albanD** [comment]
I think we can use `torch.finfo(torch.float32).eps` (replace torch.float32 with `t.dtype`.

**Comment by 1994cxy** [comment]
> Do you have ideas on a good policy for how to adjust eps based on the type?
I thought a good solution may be to judge the dtype of the denominator first and add different eps depending on the dtype. For example, I use eps=1e-6 for the float16 (it may not be the best choice of eps).
Also, other eps in the torch that intend to prevent division zero may also face the same problem, not only in the optimizer.

**Comment by 1994cxy** [comment]
> I think we can use `torch.finfo(torch.float32).eps` (replace torch.float32 with `t.dtype`.

I'm not sure with it, since torch.finfo(torch.float16).tiny=6.103515625e-05 also works for float16.
And it is smaller than torch.finfo(torch.float16).eps=0.0009765625 on my computer.
Eps may need to be small enough for a smaller impact on the result.
Also, if we limit the problem to several types of precision, we can use different default eps for them(1e-8 for float32 and 1e-6 for float16).
And for other types of precision, just print a warning.
